src/main/flight/gyroanalyse.c

```diff
@@ -311,25 +311,30 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state,
         }
         case STEP_CALC_FREQUENCIES:
         {
-            bool fftIncreased = false;
-            float dataMax = 0;
-            float dataMin = 0;
-            uint8_t binStart = 0;
-            uint8_t binMax = 0;
-            //for bins after initial decline, identify start bin and max bin 
+            // identify max bin and max/min heights 
+            float dataMax = 0.0f;
+            float dataMin = 1.0f;
+            uint8_t binMax = 0; 
             for (int i = fftStartBin; i < FFT_BIN_COUNT; i++) {
-                if (fftIncreased || (state->fftData[i] > state->fftData[i - 1])) {
-                    if (!fftIncreased) {
-                        binStart = i; // first up-step bin
-                        fftIncreased = true;
-                        dataMin = state->fftData[i - 1];
-                    }
+                if (state->fftData[i] > state->fftData[i - 1]) { // bin height increased
                     if (state->fftData[i] > dataMax) {
                         dataMax = state->fftData[i];
+                        dataMin = state->fftData[i];
                         binMax = i;  // tallest bin
                     }
                 }
             }
+            if (binMax == 0) { // no bin increase, hold prev max bin, dataMin = 1 dataMax = 0, ie move slow
+                binMax = lrintf(state->centerFreq[state->updateAxis] / fftResolution);
+            } else { // there was a max, look backwards for min, don't compare to bin 0
+                for (int i = binMax; i > 1; i--) {
+                    if (state->fftData[i] < state->fftData[i - 1]) { // up step below this one
+                        dataMin = state->fftData[i]; // set min if smaller exists before max
+                        break;
+                    }
```

Comment:
I'm not sure how often this scenario might actually occur in real life or in the bigger picture how important it actually is, but I think this logic could potentially greatly under appreciate the magnitude of noise spikes that have a 'shape' where there is a false summit followed by a little valley just before the actual peak.

Let say a the noise floor in the FFT is '1', and the FFT bucket contents have a section like this, where '9' is the value in the binMax bucket (values completely made up and aren't floats, but relative value is what is relevant to the example):
[... 1, 1, 1, 8, 7, 9, 1, 1, 1 ...]

It would look backwards starting from the binMax (the one with the value '9') -- it would stop at dataMin = 7 because of the '8', and the dataMax:dataMin ratio used to set the dynamicFactor would be a modest 9:7, which could make the notch slower to target this spike if it suddenly popped up as a nasty resonance at a certain throttle value or something.

That leaves open what a 'better' way of selecting dataMin would be without getting too computationally expensive... perhaps have it only 'break' in the if after a minimum number of steps down from binMax (and only setting dataMin if the new value would be lower than the previous).

Reply:
It is very difficult.  Originally the code just got the first minimum it came across.  This method works better. 

In the example you provide, there are two peaks.  It would go to 9 because that was biggest. The weighting would set the actual filter location half way between 7 and 9, almost in the middle of the whole 8,7,9 peak.  It would stay on 9 indefinitely, until 9 became less than some other peak, even though it would perceive 9 as being a 'weak' peak, because it would always go to the biggest.  Should peak 9 fade, and become smaller than 8, the focus would shift to the 8 peak, and the move would be fairly quick.  But the centre would be also half-way through 7, or close to the middle of the whole lump, and wouldn't shift all that much.

Should the array be something like [... 1, 3, 7, 2, 1, 5, 9, 2, 1 ...] then we have two peaks, one big and one small.  It will go for 9, and stay there until either 9 drops below 7, when it would head to the peak at 7, fairly quickly.  If these peaks are similar in shape, it may jitter from one to the other.  Widely separated and similar height peaks like will always be a problem, though the code moves relatively quickly if they are large, and will favour staying on the most persistent.  If the moves from one to the other are delayed by smoothing, it will never stay on either of them.  That's why it has to move quickly to big peaks.

If the peak at 9 was resonance, and the pattern was [... 1, 3, 2, 1, 1, 5, 9, 2, 1 ...] the notch would go to 9, and would typically reduce the resonance, perhaps making the peak at 3 the largest after about 15-20ms.  But it would move only relatively slowly down towards 3 since that is a weak peak.  Resonance being resonance, as the notch moved away from the resonant point, the resonant peak would tend to return.  As it came back, the relatively slow drift of the FFT away from the resonant point would stop, and the notch would move back towards the main resonant peak.  It that resonant peak was fairly strong, it would not move far away before the resonance tended to come back and re-centre the filter.

I tested quite a few alternatives and this method of detecting max and min gave the best results.  If anyone can figure a better way, so we can do A-B testing, I'd welcome that.

Reply:
@dogjutsu - I will test looking forward from a peak to get a 'higher-bin min', as well as looking backwards, as I do now.  We can then use the lowest height of either of those.  That might better address the issue you raise than just looking backwards.

Reply:
@ctzsnooze this sounds like a good solution to me -- should be more reliably representative of the relative amplitude of the peak.